### PR TITLE
scale conversation widgets as window resizes

### PR DESCRIPTION
**Pull request by creviera**
# Description

Fixes https://github.com/freedomofpress/securedrop-client/issues/1061
Resolves https://github.com/freedomofpress/securedrop-client/issues/1204
Known issues: https://github.com/freedomofpress/securedrop-client/issues/1205, https://github.com/freedomofpress/securedrop-client/issues/1207, https://github.com/freedomofpress/securedrop-client/issues/815, 

Known issue that this PR would introduce: https://github.com/freedomofpress/securedrop-client/issues/1213

# Test Plan

## Basic Test
1. Run the client against a server with some sources
2. Select a source and resize the client window, try double clicking and resizing to the smallest width and height
    - [ ] Confirm there is never horizontal scrollbar in the conversation view
    - [ ] Confirm that content is wrapped within message and reply speech bubbles as you resize the client window (note that https://github.com/freedomofpress/securedrop-client/issues/815 is a known issue)
## Test supported minimum screen resolutions
1. Change your screen resolution to 1280x1024
2. Run through basic testing procedure above
3. Change your screen resolution to 1368x768
4. Run through basic testing procedure above
## Test pending and failed replies
1. Close the client
2. Run this code so that you can create a pending and failed reply in the next steps: `wget https://gist.githubusercontent.com/creviera/7f19a7d10334359f40dbdbb2354cd13a/raw/a2ef94913a155aa4019b753cf916f844c9cffa3a/pending-reply && git apply pending-reply`
3. Log into the client and send a reply to a source, see it pending along with a server connection error message
4. With that source still selected, resize the client a bunch, including to its smallest size, maximize, double-click multiple times, etc.
    - [ ] Confirm there is never horizontal scrollbar in the conversation view
    - [ ] Confirm that content is wrapped within message and reply speech bubbles as you resize the client window (note that https://github.com/freedomofpress/securedrop-client/issues/815 is a known issue)
5. Log out and back in to see the pending reply show up with a failure message
6. Repeat step 4, this time with the failed reply in the conversation view
    - [ ] Confirm there is never horizontal scrollbar in the conversation view
    - [ ] Confirm that content is wrapped within message and reply speech bubbles as you resize the client window (note that https://github.com/freedomofpress/securedrop-client/issues/815 is a known issue)
## Test new messages and replies from the JI
1. Send a message as a source from the JI while the client has the conversation view for that source open
    - [ ] Confirm that the message is displayed in a speech bubble of the correct size no matter the width of the client window (should match the size of other message speech bubbles in the conversation view)
2. Send a reply from the JI to a source that is currently selected in the client you have running
    - [ ] Confirm that the reply is displayed in a speech bubble of the correct size no matter the width of the client window (should match the size of other reply speech bubbles in the conversation view)
## Test file attachments
1. Select a source with a file attachment that has not yet been downloaded
    - [ ] Confirm the download button, "ENCRYPTED FILE ON SERVER" text, and file size all fit within the file widget when the client is scaled down to its minimum width
    - [ ] Confirm the horizontal line between the text and file size does not exist when the client is scaled down to its minimum width
    - [ ] Confirm the horizontal line between the text and file size stretches horizontally as the client window is resized
2. Download the file
    - [ ] Confirm the export/ print buttons, filename, and file size all fit within the file widget when the client is scaled down to its minimum width (note https://github.com/freedomofpress/securedrop-client/issues/1205 is a known issue)
    - [ ] Confirm the horizontal line between the filename and file size does not exist when the client is scaled down to its minimum width (the filename has to be long enough to not see the horizontal line)
    - [ ] Confirm the horizontal line between the text and file size stretches horizontally as the client window is resized

# Checklist

If these changes modify code paths involving cryptography, the opening of files in VMs or network (via the RPC service) traffic, Qubes testing in the staging environment is required. For fine tuning of the graphical user interface, testing in any environment in Qubes is required. Please check as applicable:

 - [ ] I have tested these changes in the appropriate Qubes environment
 - [ ] I do not have an appropriate Qubes OS workstation set up (the reviewer will need to test these changes)
 - [ ] These changes should not need testing in Qubes

If these changes add or remove files other than client code, the AppArmor profile may need to be updated. Please check as applicable:

 - [ ] I have updated the [AppArmor profile](https://github.com/freedomofpress/securedrop-client/blob/HEAD/files/usr.bin.securedrop-client)
 - [ ] No update to the AppArmor profile is required for these changes
 - [ ] I don't know and would appreciate guidance

If these changes modify the database schema, you should include a database migration. Please check as applicable:

 - [ ] I have written a migration and upgraded a test database based on `main` and confirmed that the migration applies cleanly
 - [ ] I have written a migration but have not upgraded a test database based on `main` and would like the reviewer to do so
 - [ ] I need help writing a database migration
 - [ ] No database schema changes are needed

### CODE REVIEW

**Comment by creviera** [comment]
Note that the client cannot currently size down any smaller in width until we make it so that the source list can resize smaller and/or collapse the left pane that shows the auth widget. We can also do things like move the "Failed to send" message or show an error image with hover text, etc. The minimum width allowed in this PR is the minimum that allows us to display all of the conversation widgets, including the reply failure message, speech bubbles (with the zigzag vertical alignment effect to show messages on the left and replies on the right), journalist badge, file widget options, "ENCRYPTED FILE ON SERVER" text, and file size.

**Comment by creviera** [comment]
There's a "fitted text" issue with this PR that I don't have a workaround for yet. To help debug, I created a couple programs that simplify what the client does. You can see how nicely label text wraps when the window is resized in https://github.com/creviera/qt-examples/blob/scalable-scrollarea-with-qlabels/test.py. But when you uncomment the string `thisisalongwordwithoutspacesordashesthisisalongwordwithoutspacesordashesthisisalongwordwithoutspacesordashesthisisalongwordwithoutspacesordashesthisisalongwordwithoutspacesordashesthisisalongwordwithoutspacesordashesthisisalongwordwithoutspacesordashesthisisalongwordwithoutspacesordashesthisisalongwordwithoutspacesordashesthisisalongwordwithoutspacesordashesthisisalongwordwithoutspacesordashesthisisalongwordwithoutspacesordashesthisisalongwordwithoutspacesordashesthisisalongwordwithoutspacesordashesthisisalongwordwit💩houtspacesordashes` you'll see the qt bug we've mentioned in the past and in this PR (see #815). You'll see the label no longer wraps the text when the window is resized. 

This brings us to the solution in https://github.com/creviera/qt-examples/blob/scalable-scrollarea-with-qlabels/qtbug-89498-workaround.py, which is a simplified version of what I'm doing in this PR. However, in this PR, I'm having an issue where the qlabel termporarily grows too tall and then fits the text, which doesn't look great. The plan is to expand this program to include right and left aligned labels that don't stretch across the entire scrollarea and keep adding in more changes to try to repro this fitted text issue.

**Review by kushaldas — APPROVED** [review]
## Basic Test
1. Run the client against a server with some sources
2. Select a source and resize the client window, try double clicking and resizing to the smallest width and height
    - [x] Confirm there is never horizontal scrollbar in the conversation view
    - [x] Confirm that content is wrapped within message and reply speech bubbles as you resize the client window (note that https://github.com/freedomofpress/securedrop-client/issues/815 is a known issue)
## Test supported minimum screen resolutions
1. Change your screen resolution to 1280x1024
2. Run through basic testing procedure above
3. Change your screen resolution to 1368x768
4. Run through basic testing procedure above
## Test pending and failed replies
1. Close the client
2. Run this code so that you can create a pending and failed reply in the next steps: `wget https://gist.githubusercontent.com/creviera/7f19a7d10334359f40dbdbb2354cd13a/raw/a2ef94913a155aa4019b753cf916f844c9cffa3a/pending-reply && git apply pending-reply`
3. Log into the client and send a reply to a source, see it pending along with a server connection error message
4. With that source still selected, resize the client a bunch, including to its smallest size, maximize, double-click multiple times, etc.
    - [x] Confirm there is never horizontal scrollbar in the conversation view
    - [x] Confirm that content is wrapped within message and reply speech bubbles as you resize the client window (note that https://github.com/freedomofpress/securedrop-client/issues/815 is a known issue)
5. Log out and back in to see the pending reply show up with a failure message
6. Repeat step 4, this time with the failed reply in the conversation view
    - [x] Confirm there is never horizontal scrollbar in the conversation view
    - [ ] Confirm that content is wrapped within message and reply speech bubbles as you resize the client window (note that https://github.com/freedomofpress/securedrop-client/issues/815 is a known issue)
## Test new messages and replies from the JI
1. Send a message as a source from the JI while the client has the conversation view for that source open
    - [ ] Confirm that the message is displayed in a speech bubble of the correct size no matter the width of the client window (should match the size of other message speech bubbles in the conversation view)
2. Send a reply from the JI to a source that is currently selected in the client you have running
    - [ ] Confirm that the reply is displayed in a speech bubble of the correct size no matter the width of the client window (should match the size of other reply speech bubbles in the conversation view)
## Test file attachments
1. Select a source with a file attachment that has not yet been downloaded
    - [ ] Confirm the download button, "ENCRYPTED FILE ON SERVER" text, and file size all fit within the file widget when the client is scaled down to its minimum width
    - [ ] Confirm the horizontal line between the text and file size does not exist when the client is scaled down to its minimum width
    - [ ] Confirm the horizontal line between the text and file size stretches horizontally as the client window is resized
2. Download the file
    - [ ] Confirm the export/ print buttons, filename, and file size all fit within the file widget when the client is scaled down to its minimum width (note https://github.com/freedomofpress/securedrop-client/issues/1205 is a known issue)
    - [ ] Confirm the horizontal line between the filename and file size does not exist when the client is scaled down to its minimum width (the filename has to be long enough to not see the horizontal line)
    - [ ] Confirm the horizontal line between the text and file size stretches horizontally as the client window is resized

File: securedrop_client/gui/widgets.py [hunk 1]
```diff
@@ -2244,34 +2287,45 @@ def __init__(
...
 
     def eventFilter(self, obj, event):
         t = event.type()
         if t == QEvent.MouseButtonPress:
             if event.button() == Qt.LeftButton:
                 self._on_left_click()
-        # See https://github.com/freedomofpress/securedrop-client/issues/835
-        # for context on code below.
-        if t == QEvent.HoverEnter and not self.downloading:
+        elif t == QEvent.HoverEnter and not self.downloading:
```
Comment (emkll) [inline securedrop_client/gui/widgets.py:2315]: nice catch

File: securedrop_client/gui/widgets.py [hunk 2]
```diff
@@ -2244,34 +2287,45 @@ def __init__(
...
             self.download_button.setIcon(load_icon("download_file_hover.svg"))
         elif t == QEvent.HoverLeave and not self.downloading:
             self.download_button.setIcon(load_icon("download_file.svg"))
+
         return QObject.event(obj, event)
 
+    def update_file_size(self):
```
Comment (emkll) [inline securedrop_client/gui/widgets.py:2322]: Looks like this was just moved down, but the logic remains unchanged.

**Review by emkll — APPROVED** [review]
Went through a subset of the very comprehensive test plan in Qubes in an sd-app VM, using a deb built from this branch. Outside of the test plan, and documented known issues the behavior feels very smooth and polished, thanks for your attention to detail here.
 
## Basic Test
1. Run the client against a server with some sources
2. Select a source and resize the client window, try double clicking and resizing to the smallest width and height
    - [x] Confirm there is never horizontal scrollbar in the conversation view
    - [x] Confirm that content is wrapped within message and reply speech bubbles as you resize the client window (note that https://github.com/freedomofpress/securedrop-client/issues/815 is a known issue)
## Test supported minimum screen resolutions ( I can't change this on my laptop Qubes :( )
1. Change your screen resolution to 1280x1024
2. Run through basic testing procedure above
3. Change your screen resolution to 1368x768
4. Run through basic testing procedure above
## Test pending and failed replies
1. Close the client
2. Run this code so that you can create a pending and failed reply in the next steps: `wget https://gist.githubusercontent.com/creviera/7f19a7d10334359f40dbdbb2354cd13a/raw/a2ef94913a155aa4019b753cf916f844c9cffa3a/pending-reply && git apply pending-reply`
3. Log into the client and send a reply to a source, see it pending along with a server connection error message
4. With that source still selected, resize the client a bunch, including to its smallest size, maximize, double-click multiple times, etc.
    - [x] Confirm there is never horizontal scrollbar in the conversation view
    - [x] Confirm that content is wrapped within message and reply speech bubbles as you resize the client window (note that https://github.com/freedomofpress/securedrop-client/issues/815 is a known issue)
5. Log out and back in to see the pending reply show up with a failure message
6. Repeat step 4, this time with the failed reply in the conversation view
    - [x] Confirm there is never horizontal scrollbar in the conversation view
    - [x] Confirm that content is wrapped within message and reply speech bubbles as you resize the client window (note that https://github.com/freedomofpress/securedrop-client/issues/815 is a known issue)
## Test new messages and replies from the JI (Did not test)
1. Send a message as a source from the JI while the client has the conversation view for that source open
    - [ ] Confirm that the message is displayed in a speech bubble of the correct size no matter the width of the client window (should match the size of other message speech bubbles in the conversation view)
2. Send a reply from the JI to a source that is currently selected in the client you have running
    - [ ] Confirm that the reply is displayed in a speech bubble of the correct size no matter the width of the client window (should match the size of other reply speech bubbles in the conversation view)
## Test file attachments
1. Select a source with a file attachment that has not yet been downloaded
    - [x] Confirm the download button, "ENCRYPTED FILE ON SERVER" text, and file size all fit within the file widget when the client is scaled down to its minimum width
    - [x] Confirm the horizontal line between the text and file size does not exist when the client is scaled down to its minimum width
    - [x] Confirm the horizontal line between the text and file size stretches horizontally as the client window is resized
2. Download the file
    - [X] Confirm the export/ print buttons, filename, and file size all fit within the file widget when the client is scaled down to its minimum width (note https://github.com/freedomofpress/securedrop-client/issues/1205 is a known issue) NOTE: I did observe item 3 of #1205
    - [X] Confirm the horizontal line between the filename and file size does not exist when the client is scaled down to its minimum width (the filename has to be long enough to not see the horizontal line)
    - [x] Confirm the horizontal line between the text and file size stretches horizontally as the client window is resized

Given my review above, and the comprehensive review provided by @kushaldas this looks good to merge.

**Pull request merged**